src/v/cloud_storage/cache_service.cc

```diff
@@ -796,6 +804,43 @@ ss::future<> cache::do_reserve_space(size_t bytes) {
     _reserved_cache_size += bytes;
 }
 
+void cache::set_block_puts(bool block_puts) {
+    if (_block_puts && !block_puts) {
+        _block_puts_cond.signal();
+    }
+    _block_puts = block_puts;
+}
+
+void cache::notify_disk_status(
+  [[maybe_unused]] uint64_t total_space,
+  [[maybe_unused]] uint64_t free_space,
+  storage::disk_space_alert alert) {
+    vassert(ss::this_shard_id() == 0, "Called on wrong shard");
+
+    bool block_puts = (alert == storage::disk_space_alert::degraded);
+
+    // Log on transitions
+    if (block_puts && !_block_puts) {
+        vlog(
+          cst_log.warn,
+          "Tiered storage cache blocking segment promotions, disk space is "
+          "critically low.");
+    } else if (!block_puts && _block_puts) {
+        vlog(
+          cst_log.info,
+          "Tiered storage cache un-blocking promotions, disk space is no "
+          "longer critical.");
+    }
+
+    // Propagate to other shards on changes
+    if (block_puts != _block_puts) {
+        ssx::spawn_with_gate(_gate, [this, block_puts]() {
+            return container().invoke_on_all(
+              [block_puts](cache& c) { c.set_block_puts(block_puts); });
+        });
```

Comment:
is it possible for the caller, which looks like upcalls from storage layer, to become flappy and dispatch disk status calls here resulting in multiple in flight background fibers? i guess i'd wonder what would happen if those were reordered and we blocked puts even tho unblocking puts had been the latest upcall.

Reply:
Disk updates are pretty far apart in walltime, so I think this is super unlikely in practice.

Maybe when we change the interface for disk notifications to include the feedback on usage etc, we can change these upcalls to be async, so that they don't need to background any cross-shard work.